### PR TITLE
Verify and report per-key quota

### PR DESCRIPTION
**Pull request by arielshaqed**
* Verify quota on usage mutations
* Add a webserver (for reporting quota exceeded)
* Add a per-key quota exceeded endpoint

Hand-assemble the HTTP server paths and JSON responses -- this is a simple GET.

## Future plans (_not_ in this PR, but possible with it once/if we need them):

- Allow long-poll quota exceeded endpoint by setting a query parameter
- Interrupt the long poll and report shortly after a key exceeds quota.

This will allow us the best of both worlds: long-polling to simplify management server operations, and immediately reporting quota violations.

### CODE REVIEW

File: cmd/terminus/main.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+func GetFlagBytes(flags *pflag.FlagSet, flag string) (int64, error) {
+	s, err := flags.GetString(flag)
+	if err != nil {
+		return 0, fmt.Errorf("get flag %s: %w", flag, err)
+	}
+	bytes, err := humanize.ParseBytes(s)
+	if err != nil {
+		return 0, fmt.Errorf("parse flag %s value %s: %w", flag, s)
+	}
+	if int64(bytes) < 0 {
```
Comment (itai-david) [inline cmd/terminus/main.go:64]: Question - Is that a reasonable limit to use? Shouldn't we use a reasonable max quota here?

**Review by itai-david — COMMENTED** [review]
2 Comments from https://github.com/treeverse/terminus/pull/5
Other than that another comment/question hereafter.
Other than that, looks excellent

File: cmd/terminus/main.go (hunk 1 above)
Comment (arielshaqed) [reply cmd/terminus/main.go:64]: There is no max on S3.  2^63 bytes is 8 EiB or 8192 PiB.  So we won't need that much in the foreseeable future.

Going beyond that, this code is anyway broken at the 63/64 bit limit.  E.g. we hold usage as a PostgreSQL BIGINT, which _is_ `int64`.  We'd need to represent bigger ints across the board -- out of scope.

File: cmd/terminus/main.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+var runCmd = &cobra.Command{
+	Use:     "run",
+	Short:   "Start the Terminus server",
+	Example: "terminus run --sqs-name=terminus-queue --db-dsn=postgres:/// --default-quota=1G",
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := context.Background()
+		dbDriver, err := cmd.Flags().GetString("db-driver")
+		DieOnErr(err)
+		dbDSN, err := cmd.Flags().GetString("db-dsn")
+		DieOnErr(err)
```
Comment (arielshaqed) [inline cmd/terminus/main.go]: You wrote on the #5 review::

> This pattern returns multiple times. Maybe we should refactor into designated function:
> ```go
> func GetStringFromFlagsOrDie(flag string) string {
>   val, err := cmd.Flags().GetString(flag)
>   DieOnErr(err)
>   return val
> }
> ```
> 
> And then:
> ```go
> dbDriver := GetStringFromFlagsOrDie("db-driver")
> dbDsn := GetStringFromFlagsOrDie("db-dsn")
> ```
> etc

I don't know what is more Go-ish (generally things-that-die are not done, but this is a UI cmd so things are different), and @nopcoder can weigh in.  So doing it your way to speed approval.

File: cmd/terminus/main.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	runCmd.Flags().StringP("sqs-name", "q", "", "Name of topic on SQS with S3 events to process")
+	runCmd.MarkFlagRequired("sqs-name")
+
+	runCmd.Flags().StringP("default-quota", "Q", "5KB", "Default quota size")
+
+	runCmd.Flags().String("db-driver", "pgx", "Database driver code")
+	runCmd.Flags().StringP("db-dsn", "d", "", "DSN to connect to database")
+	runCmd.MarkFlagRequired("db-dsn")
+
+	runCmd.Flags().StringP("pattern", "p", `^s3://[^/]*/user/([^/]*)/.*$`, "Regexp matching paths to track")
```
Comment (arielshaqed) [inline cmd/terminus/main.go]: You wrote:

> ```suggestion
> 	runCmd.Flags().StringP("pattern", "p", `^s3://[^/]+/user/([^/]+)/.*$`, "Regexp matching paths to track")
> ```
> I might be totally wrong, but it seems the original allows empty sequnces between `/` except for the 'user' literal. Is that on purpose?

Yes it does.  You care where I don't, and it's a flag that can *and will* be changed by users, so made this change!

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Closing #5 and continuing here -- I mixed up the rebases and bases, and it will be easier to put the two together.  Sorry.

**Review by itai-david — APPROVED** [review]
Thanks for the changes

**Comment by arielshaqed** [comment]
Thanks!

Pulling...

**Pull request merged**